[1단계 - 장기] 체체(김진영) 미션 제출합니다.

CheChe903 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)
-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4117

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

생활체조들을 지킨 것도, 지키지 못한 것이 있습니다.
<!-- 이유 작성 -->

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 찰리~! 7기 체체입니다. 이번 리뷰 잘부탁드립니다. 
구현이 너무 오래 걸려 썩 좋지 않은 코드입니다. 
해당 부분들로 피드백을 받는다면 아까울 것 같아서, 현재로서는 안해주셔도 됩니다! ( 다음 피드백 요청때 반영하겠습니다.)

- [x] 디렉션 이넘 또는 객체를 만들어 기물 경로 생성 중복 제거 ( 아직 방법은 떠오르지 않지만 시도 해보겠습니다!)
현재는 중복 코드들로 각각 기물들(마, 차, 상 등등)이 반복되는 구조로 가능한 경로를 생성하고 있습니다.

- [x] 클린코드 indent 무시

- [ ] Controller 책임을 Board로 이전 
현재는 게임 진행 흐름을 컨트롤러가 담당하는 느낌입니다. 책임을 이전시켜보도록 하겠습니다.

- [x] 장기판 구성 4가지 입력 및 초기화 로직
원래는 외상차림, 내상차림, 바깥상차림, 안상차림 이렇게 4가지의 장기판 구조가 있는데, 이를 적용시켜보도록 하겠습니다!

- [x]  코드 컨벤션 부분
다음 리뷰 요청 때 모두 지켜서 제출해보도록 하겠습니다.

원하는 리뷰는 캡슐화가 잘 지켜졌나? 입니다.

궁금한 점

1. 현재 Board에서 해당 기물이 차인지, 포인지 확인하는 로직이 존재합니다. 
크루원 한명이 말해주길, getClass 또는 instanceOf를 사용하는 것은 추상 클래스를 사용한 의미가 없다. 라고 해주었는데요, 
이러한 상황이 처음인 만큼 저만의 기준을 잡기가 어렵습니다. 해당 부분을 찰리가 저만의 기준을 잡을 수 있도록 도와주실 수 있을까요 ?



<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

CheChe903 committed:
docs(README): 1.1단계 요구사항 작성

CheChe903 committed:
feat(Position): 위치 동등성 로직 및 단위 테스트 구현

CheChe903 committed:
feat(Position): 위치 범위 검증 로직 및 단위테스트 구현

CheChe903 committed:
feat(Piece): 기물 움직임 비즈니스 로직 및 단위 테스트 구현

CheChe903 committed:
feat(Piece): 기본 장기 말 클래스 구현

CheChe903 committed:
feat(Piece): 턴 및 팀 사이즈 검증 구현

CheChe903 committed:
feat(Turn): 턴 정적 팩토리 메서드 구현 및 턴 교체 비즈니스 로직, 단위 테스트 구현

CheChe903 committed:
refactor(PositionTest): 메서드 네이밍 변경

CheChe903 committed:
docs(README): 요구사항 업데이트

CheChe903 committed:
feat(PiecesInitializer): 장기 말 초기화 로직 및 단위 테스트 구현

CheChe903 committed:
feat(Pieces, Board): 보드 및 장기 말들 초기화 로직 구현

CheChe903 committed:
docs(README): 요구사항 업데이트

CheChe903 committed:
feat(Route): 경로 점유 및 단위 테스트 구현

CheChe903 committed:
feat(Route): 루트 동등성 체크 로직 및 단위 테스트 구현

CheChe903 committed:
feat(Soldier): 가능한 경로 체크 비즈니스 로직 및 단위 테스트 구현

- Route를 저장하는 방식도 List에서 Set으로 변경합니다.

CheChe903 committed:
feat(Guard): 가능한 경로 체크 비즈니스 로직 및 단위 테스트 구현

CheChe903 committed:
refactor: 경로 구조 수정

- RawRoute 생성으로 해당 루트가 유효한 지 파악 후 유효하다면 진행합니다.

CheChe903 committed:
feat(Piece): 기물 이동 및 단위 테스트 구현

CheChe903 committed:
feat(Pieces): 기물에 따라(포, 차, 그 외) 가능한 경로 계산 비즈니스 로직 및 단위 테스트 구현

CheChe903 committed:
feat(Piece): 해당 피스가 입력한 팀과 같은 팀인지 판단하는 비즈니스 로직 및 단위 테스트 구현

CheChe903 committed:
refactor(Turn): 턴 받기, 턴 교체 메서드 분리

CheChe903 committed:
feat(Board): 기물 선택, 경로 탐색, 움직임 로직 및 단위 테스트 구현

CheChe903 committed:
fix(Pieces): 특정 기물 움직임 버그 수정

CheChe903 committed:
fix(Pieces): 차 기물 이동 버그 수정

CheChe903 committed:
feat: 게임 진행

CheChe903 committed:
docs(README): 요구사항 업데이트

CheChe903 committed:
refactor(PiecesTest): 로직에 맞게 테스트 수정

CheChe903 committed:
refactor(Piece): 메서드 분리 및 컨벤션

CheChe903 committed:
refactor(Pieces): 메서드 분리 및 컨벤션

CheChe903 committed:
refactor(Horse): 메서드 접근 제어자 변경

CheChe903 committed:
refactor: final 키워드 등록

CheChe903 committed:
refactor: final 키워드 등록

CheChe903 committed:
refactor(Elephant): 메서드 접근 제어자 변경

CheChe903 committed:
refactor(Piece): 추상 메서드 위치 컨벤션

Gomding commented on README.md:
```diff
@@ -1,3 +1,118 @@
+
+- [x] 위치는 x축 기준 0-8, y축 기준 0-9만 가능하다.
+
+## 기물 모음
```
요구사항을 잘 작성해주셨지만
기물 관련해서는 아직 빈약한 것 같아요 😢 

> 구현 전에 해야 할 일
> 
> 개발을 시작하기 전에 어떤 기능을 구현해야 하는지 정리하라.
> 정리한 내용을 .md 파일로 제출하라.

1.1 단계의 과제 진행 요구사항을 참고했을 때
기물 모음을 읽고 장기 도메인에서 기물들을 어떻게 구현해야할지 알 수 있을까요?
장기에서 중요한건 여럿 있지만 `각 기물이 어떻게 움직이냐` 라는건 중요한 규칙이라 생각해요 😃 
체체는 어떻게 생각하시나요?

CheChe903 replied:
`기물이 어떻게 움직이냐` 라는 규칙은 통상적으로 모두 아는 개념이라고 착각하여 적지 않았습니다. 
하지만 장기를 모르는 사람이 볼 수 있는 코드이기 때문에 추가하는 게 맞다고 생각합니다!

이제는 제가 아는 규칙들도 적어보려고 노력해보겠습니다!

[37e2f8e](https://github.com/woowacourse/java-janggi/pull/63/commits/37e2f8e628be23f8517d24bb7b21ff771ee0eb2d)

Gomding replied:
말씀하신대로 익숙한 규칙이 누군가는 익숙하지 않을 수 있어요 😃 
또한 미션 요구사항을 참고하면

> 단순히 코드를 작성하는 것이 아니라, 무엇을 개발해야 하는지 먼저 정리하고 개발을 진행하는 것이 핵심이다. 

요구사항이란 무엇을 개발해야하는지 정리하는게 목적이므로
현재 모르고있는 개념만 적는게 아닌 현재 구현해야하는 도메인에 대해 서술한 것이라 생각해요.
때문에 내용은 친절 할 수록 좋다고 생각합니다 :)

누구나 아는 지식이겠지 하고 누락했다가 안좋은 과정을 겪은적이 많습니다 ㅎㅎ..

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -0,0 +1,46 @@
+            final Team currentTurn = board.getTurn();
+            outputView.printTurn(currentTurn);
+
+            final Position position = inputView.inputPiecePosition();
```
잘못 입력했더니 프로그램이 종료되는데 해당 부분은 다시 입력할 수 있게 한다면 좋겠어요!

![image](https://github.com/user-attachments/assets/7ea30f31-cbb5-4b61-a720-95be9a64eb70)

CheChe903 replied:
해당 부분은 아직 요구사항에 있지 않아 완성시키지 않았어요!

2.1 단계에서 이루려고 했던 요구사항입니다.

Gomding replied:
넵 해당 부분은 2단계에 적용해주셔도 좋습니다! 😃 

중요한 것은 아니지만 다음 관점에서 있으면 좋겠다고 생각했어요!

1. 사용자의 불편함 (한 번 입력 삐끗하면 장기 처음부터 다시 해야함..)
2. 리뷰하는 과정에서 실행 후 테스트하는 경우도 많음

CheChe903 replied:
그렇네요, 2번이 좀 와닿는 것 같아요 ! 
리뷰는 코드만 보는 것이 아닌 실행도 포함되니 리뷰어분들이 편하게 리뷰할 수 있는 환경을 조성하는 것이 좋을 것 같아요.

이번 단계에 적용해볼게요!

[31b9d53](https://github.com/woowacourse/java-janggi/pull/63/commits/31b9d53773b659d4d16a84f950b94c3b69e86944)

Gomding commented on src/main/java/janggi/domain/Turn.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.List;
+
+public class Turn {
+    private final List<Team> teams;
```
턴에 대한 내부를 리스트로 구현하신 이유가 궁금해요!
더 심플하게 구현하는 방법은 없었을까요? 😃

CheChe903 replied:
심플하게 team 하나만 넣는 방법도 있군요 !
페어와 이야기하며 rotate를 사용하자는 말에 매몰되어 List로 작성했어요.

[fced947](https://github.com/woowacourse/java-janggi/pull/63/commits/fced94781c18369c8c2c7868935c4b246a72c228)

Gomding replied:
물론 팀이 셋 이상이라면 좋은 구현이라 생각해요! 👍 
현재는 더 간단한 방법이 있어 드린 의견이었습니다 :)

Gomding commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,57 @@
+        super(position, team);
+    }
+
+    @Override
+    public boolean isCannon() {
+        return true;
+    }
```
> 현재 Board에서 해당 기물이 차인지, 포인지 확인하는 로직이 존재합니다.
> 크루원 한명이 말해주길, getClass 또는 instanceOf를 사용하는 것은 추상 클래스를 사용한 의미가 없다. 라고 해주었는데요,
> 이러한 상황이 처음인 만큼 저만의 기준을 잡기가 어렵습니다. 해당 부분을 찰리가 저만의 기준을 잡을 수 있도록 도와주실 수 있을까요 ?

해당 부분은 1차적으로 고민해야할 부분은 
차인지 포인지 확인해야하는 과정이 꼭 필요한가를 고민해봐야할 것 같아요 😃 
어떤 문제를 해결하기 위해 차인지 포인지 확인하게 됐을까요?
현재 이동에 대해서는 각 기물이 책임을 가지고 있는데 기물이 스스로 판단할 수 없는 상황이었는지 궁금해요 🤔 

---

getClass 와 instanceOf 를 사용하는건 어째서 추상클래스를 사용한 의미가 없다고 하는걸까요? 😃 
체체의 의견을 들어보고 싶어요!

CheChe903 replied:
>해당 부분은 1차적으로 고민해야할 부분은
차인지 포인지 확인해야하는 과정이 꼭 필요한가를 고민해봐야할 것 같아요 😃
어떤 문제를 해결하기 위해 차인지 포인지 확인하게 됐을까요?
현재 이동에 대해서는 각 기물이 책임을 가지고 있는데 기물이 스스로 판단할 수 없는 상황이었는지 궁금해요 🤔

왜 기물이 무엇인지 알아야하나 ?
-> 이동 가능한 경로를 생성하는 부분에서 차, 포, 그 외 기물의 방식이 모두 규칙이 다릅니다. 따라서 `Pieces`에서 기물에 따라 불러지는 메서드가 다른데요. 그래서 어떠한 기물인지 판단하는 로직이 `Board`에 필요하게 되었어요.

먼저, 간단하게 해결할 수 있는 방법은 `Piece`가 `Pieces`를 알게되면 해결할 수 있어요. 
현재 코드는 `Piece`에서 독립적으로 갈 수 있는 경로를 생성(다른 기물의 위치와 상관없이) -> `Pieces`에서 다른 기물들과의 관계를 파악하여 최종적으로 갈 수 있는 루트를 생성합니다!

따라서 Piece 혼자서 독립적으로 가능한 확정된 경로를 파악할 수 없기 때문에 무슨 기물인지 알아야합니다. 
Piece가 스스로 다른 Piece를 생각하여 (Piece의 calculateRoute() 의 파라미터에 Pieces 주입) 경로를 만든다면 기물을 몰라도 됩니다!

하지만 이렇게 진행한다면 순환 참조이기 때문에

----
라고 적는 와중에 Piece가 List<Piece>를 아는 것은 순환 참조가 아닌 것 같아서 작성을 멈추고 리팩토링을 해보았어요.
적는 와중에 생각이 정리되었습니다.

TODO 커밋로그

CheChe903 replied:
해당과 관련된 제 의견을 적다가 리팩토링하였습니다!

기존에는 `Piece`가 독립적으로 갈 수 있는 경로를 계산 -> `Pieces`에서 갈 수 있는 경로 중, 다른 기물에 의해 못가는 경우 제외 후 반환
이었습니다.
이렇게 하지 않는다면 `Piece`가 `Pieces`를 알아야합니다. 따라서 순환 참조가 일어날 것으로 예측하고 설계를 본래와 같이 진행했습니다.

 하지만 이에 대한 제 의견을 적다보니, `Piece`가 `List<Piece>`를 아는 것은 전혀 순환 참조가 아니라고 판단되었어요!
 
 그래서 기물 확인 과정을 삭제하는 리팩토링을 진행했습니다.
[9c6257f](https://github.com/woowacourse/java-janggi/pull/63/commits/9c6257f1e0e3a3404f00ea52d5566813e407bd1e)
 
>`getClass` 와 `instanceOf` 를 사용하는건 어째서 추상클래스를 사용한 의미가 없다고 하는걸까요? 😃

다형성 파괴
결국에는 구현체들을 감추기 위해 추상클래스를 사용한다고 생각합니다. 근데 외부에서 구현체를 알아야하는 로직이 있다면 차라리 따로 모두 만들어주는 것이 더 객체지향적이라고 생각합니다.

`OCP` 위반
새롭게 또 확인해야하는 기물이 생긴다면 if문이 지속적으로 늘어납니다. 따라서 기본 클래스와 다를 게 없다고 생각합니다.

Gomding commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,57 @@
+    protected Set<RawRoute> calculateRawRoutes() {
+        final Set<RawRoute> rawRoutes = new HashSet<>();
+
+        for (int i = position.x() + 1; i <= 8; i++) {
```
여기 for 문의 종료 조건을 위한 8,0,9,0 기물이 알아야하는 정보일까요?

CheChe903 replied:
삭제되었습니다! 
하지만 해당 부분에도 0,8,9가 남아있어 `BoardSize`를 이용했습니다.

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -0,0 +1,46 @@
+            final Piece selectPiece = board.selectPiece(position);
+
+            final Set<Route> possibleRoutes = board.findPossibleRoutes(selectPiece);
+            outputView.printPossibleRoutes(possibleRoutes);
```
기물의 이동 가능 경로를 보여주는것은 매우 친절해보이네요 🥹

![image](https://github.com/user-attachments/assets/785498c2-e4c9-4a63-9d3f-c4939be7ec0d)

CheChe903 replied:
감사합니다 ! 이 기능이 저희 페어 팀의 특별한 기능입니다 ! ☺️
하지만 아직 정렬이 안되어 있어 정렬하였습니다!

Gomding commented on src/main/java/janggi/domain/piece/Horse.java:
```diff
@@ -0,0 +1,72 @@
+        return false;
+    }
+
+    private List<RawPosition> makeRawPositions1() {
```
`makeRawPositions~~` 메서드 이름이 좋지 않은것 같네요 😭

CheChe903 replied:
수정하며 사라졌습니다!

Gomding commented on src/main/java/janggi/domain/Board.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.List;
+import java.util.Set;
+
+public class Board {
```
Board 에 대한 테스트는 따로 작성하지 않으신 이유가 있을까요?

CheChe903 replied:
제 페어는 최상위 객체(Board)의 테스트는 하지 않는다고 하였습니다.
저는 보통 모두 하는 편인데, 페어의 의견에 따라 진행하지 않은 후, 시간이 남으면 진행하려했지만 시간이 많지 않아 작성하지 않았습니다 ㅠㅠ

[ef89ffd](https://github.com/woowacourse/java-janggi/pull/63/commits/ef89ffdf6bb7a7d4fe4cc0231adb1b78a88a8e87)

Gomding replied:
체체는 최상위 객체의 테스트를 필요로하는 이유도 궁금하군요! 🙋

CheChe903 replied:
제가 생각하는 테스트의 의미는 '어느 부분에서 문제가 발생하느냐?'를 판단하는 기준입니다.
먼저, `Board` 에서 문제가 생겼다고 가정하겠습니다!
개발자가 스스로 실행(`Application` 실행)하며 테스트를 진행, 기능이 의도한 바대로 되지 않을 때 보통 테스트를 모두 돌려볼 것입니다. `Board`에 대한 테스트가 없어 테스트가 모두 성공한다면, 문제점을 찾는데 시간이 더 걸릴 것 같아요.

결론적으로 
`모든 객체 테스트 작성 시간` < `오류 발생 시 테스트를 작성 하지 않은 곳의 오류 찾는 시간`
이라고 생각합니다!
또한 의도한대로 되지 않음에도 모든 테스트가 통과하니, 통과하는 테스트를 의심하는 상황도 발생할 수 있을 것 같아요.

Gomding commented on src/test/java/janggi/domain/piece/CannonTest.java:
```diff
@@ -0,0 +1,71 @@
+
+class CannonTest {
+
+    @DisplayName("차 기물이 올바른 루트를 계산하는지 확인한다.")
+    @Test
+    void calculateRoutesTest() {
```
테스트는 문서의 역할도 한다고 생각해요

테스트를 더 세분화 했으면 좋겠어요!
테스트를 읽는것만으로 각 기물의 이동에 대해 이해 할 수 있는정도로요 😃

CheChe903 replied:
찰리, 이 부분은 잘모르겠어요!

3월 22일 토요일 놀면뭐하니에서, 체스를 최종 코딩테스트처럼 구현하는 경험을 했어요. 이때 네오가 준 스켈레톤 코드가 존재하는데 네오가 쉽게 작성했음에도 불구하고 이해하는데 시간이 좀 걸리더라구요! 테스트를 중점으로 보며 좀 더 이해하기 수월했던 것 같아요.

현재 제 기물 이동은 반복문을 통해 만들고 있어요. 이 부분을 다른 사람이 처음 보면 이해가 잘 가지 않을 것 같아요. 
제가 이해한 찰리의 말은 테스트 코드가 좀 더 직관적이면 좋다는 걸로 이해했는데, 맞을까요?

Gomding replied:
다음의 사항들에 필요성을 느꼈어요!

1. `차 기물이 올바른 루트를 계산하는지 확인한다.` 라는 DisplayName을 봤을 때 `차에게 올바른 루트란 무엇인가?` 라는 의문이 생겨요!
2. 차 기물의 이동 규칙은 4방향이 있는데 하나의 테스트로 커버될 수 있을까? 라는 의문이 생겼어요!

> 테스트는 문서의 역할도 한다고 생각해요

요걸 다른말로 하면 테스트 코드를 봤을 때 
`요구사항 문서와 동일하게 `객체가 어떤 행위를 하는지 명확하게 이해할 수 있어야한다` 입니다
테스트는 실행 가능한 명세서(spec) 이라고도 해요 :)

다음 사항이 들어가면 좋다고 생각합니다 (항상 모두 적용할 필요는 없습니다 ㅎㅎ)
- 어떻게 사용할 수 있는지 보여주는 코드 예시
- 무슨 조건에서 어떤 동작을 해야하는지를 설멍
- 실제 구현이 명세 (요구사항)을 지키고 있는지 확인
  - 예외 사항도 확인

> - [x] 차
>    - [x] 차 기물은 상, 하, 좌, 우 모두 원하는 만큼 움직일 수 있다.

차 기물에 대한 요구사항을 이렇게 정리했으니
참고해서 `DisplayName` 을 사용해보는것도 좋구요 :)

CheChe903 replied:
전체적으로 점검 후 리팩토링 해보겠습니다 !

[edd615f](https://github.com/woowacourse/java-janggi/pull/63/commits/edd615fbcec42677e53c2ea5e003b32d7512e2b0)

Gomding commented on src/main/java/janggi/domain/piece/General.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+import java.util.Set;
+
+public class General extends Piece {
```
장기 도메인에서 사용하는 이름을 활용한 점이 좋네요!! 👍 👍 👍

CheChe903 replied:
감사합니다 !

Gomding commented on src/main/java/janggi/domain/Board.java:
```diff
@@ -0,0 +1,50 @@
+
+public class Board {
+
+    private final Pieces pieces;
```
Board 가 포지션을 관리하는 방법도 있었을 것 같은데
기물만 가지도록 한 배경도 궁금해요!

CheChe903 replied:
Map<Position, Piece>을 사용하는 방법도 있지만, 이 방법은 객체지향을 깨고 절차지향을 사용한다고 생각합니다!
그리고 현재 학습하는 입장에서 Map으로 편하게 구현한다면 공부하는 의미도 없다고 생각해요.

Map으로 구성된 것은 어떻게든 객체 안의 인스턴스 필드로 갖자는 소신을 따른 결과입니다!

Gomding replied:
좋은 기준을 가지고 구현해주셨네요! 💯 💯 

Board가 포지션을 관리하는 방법에 대해서
아래의 관점에서도 의견을 나눠보면 좋을 것 같아요!
(절대 아래 방법이 좋다는 관점은 아닙니다 ㅎㅎ 체체와 함께 트레이드 오프를 고민해보고 싶네요)

- Board 는 공간을 관리하고
- Piece 는 기물의 이동을 관리한다

CheChe903 replied:
> `Board`는 공간을 관리한다와 `Piece`는 기물의 이동을 관리한다.

제가 만약 이러한 관점을 갖고 개발을 한다면, `Map<Position, Piece>`가 들어가야 할 것 같아요. 

그렇다면 `Pieces`에서 해당 위치에 있는 `Piece`를 찾는`find` 메서드는 삭제되고, `Board`에서 손쉽게 찾을 수 있게 됩니다.
그런 다음, 해당 `Piece` 에게 이동을 하라고 메시지를 던지면 이동에 대한 프로덕트 코드가 완성될 것 같아요.

결국에 두 차이는 `원하는 위치의 기물을 찾는 책임` 을 `Board`가 수행하냐, `Pieces`가 수행하냐의 차이입니다.

첫번째로, 응집도에 의해 `Piece`가 해당 책임을 갖는 것이 좋다고 판단하였습니다.
`Board`는 게임을 위한 판이긴 하지만, 사실상 하나의 `Game`과 비슷한 책임을 갖고 있어요. 따라서 각 기물이 스스로 위치를 알고 움직이는 것이 더 올바른 객체라고 판단되었습니다.

두번째, `DB`의 저장, 불러오기가 쉬운가? 

게임 자체를 저장해야하는 다음 단계 미션에는 `Map<Position, Piece>`를 관리하기 힘들 것 같아요. 
하나의 `Board`에는 `Pieces`, `turn`만 있고,
각 `Piece`에는  자신의 `Position`과 `team` 등이 관리된다면 
데이터 저장도 쉽고, 데이터 불러오기가 쉬울 것 같아요.

하지만 `Map`으로 저장된다면 이를 **변환하는 로직**이 또 필요하게 됩니다. (`Map -> 데이터 저장`,` 저장된 데이터 -> Map`)

Gomding replied:
좋은 의견 감사해요 💯 💯 💯 
훌륭한 기준이라 생각합니다 ㅎㅎ

`DB 의 저장, 불러오기가 쉽다` 는 부수적인 효과라고 생각해도 될까요?
객체의 설계에 있어서 DB 에 관한 부분을 고려한다면 DB 에 의존적인 설계가 될 것 같아요
물론 프로그램은 보통 저장소(DB)에 데이터를 저장하여 사용하기에 아예 고려를 안할 수는 없지만
`객체 설계 >>>> DB 고려` 의 중요도라고 생각해요.

체체는 어떻게 생각하시나요?

Gomding reviewed (changes requested):
안녕하세요 체체!
장기 미션 함께하게된 찰리입니다 🙇 

1단계 미션 진행하느라 고생하셨어요!
몇가지 코멘트 남겼으니 확인부탁드려요 :)
궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요~

CheChe903 committed:
refactor: 패키지 구조 변경

CheChe903 committed:
refactor: 기물 이동 로직 변경

- 기존의 RawRoute, RawPosition을 삭제합니다.
- 또한 모든 기물의 이동 경로를 Direction을 이용하여 재구성합니다.

CheChe903 committed:
refactor(Direction): 미사용 메서드 제거

CheChe903 committed:
refactor(Piece): 추상 메서드 시그니쳐 변경

CheChe903 committed:
refactor(Piece): getClass 사용 삭제

CheChe903 committed:
refactor: final 키워드 추가

CheChe903 committed:
refactor: final 키워드 추가 및 공백 컨벤션

CheChe903 committed:
feat(BoardSetup): 상차림 추가

CheChe903 committed:
refactor: 테스트 수정 및 단위 테스트 추가

CheChe903 committed:
docs(README): 요구사항 업데이트

CheChe903 committed:
refactor(Turn): 턴 구현 방식 변경

- 기존에는 List<Team>를 인스턴스 변수로 갖고 있었지만 이제는 team 변수 하나만 갖고 final을 제거하여 team을 변경하는 방식으로 진행합니다.

CheChe903 committed:
refactor(Piece): Piece가 경로 구할 때 직접 List<Piece>를 받도록 변경

- 기존에는 Pieces에서 이동가능한 경로를 한 번 더 구현해야했습니다. 이를 해결하고자 Piece가 Piece List를 받아 직접 해결합니다.

CheChe903 committed:
refactor: 사용하지 않는 메서드 제거

CheChe903 committed:
refactor(Cannon, Chariot): 보드 사이즈 주입

CheChe903 committed:
feat(OutputView): 갈 수 있는 경로 정렬

CheChe903 committed:
test(BoardTest): 보드 단위 테스트 작성

CheChe903 committed:
test(BoardSizeTest): 보드 사이즈 검증 테스트 구현

- 메서드 시그니처도 validate 가 아닌 isInBoard로 변경합니다.

CheChe903 committed:
refactor: 테스트 패키지 구조 변경

CheChe903 committed:
refactor: 이동 경로 로직 수정

CheChe903 committed:
refactor: 사용하지 않는 메서드 삭제

Gomding commented on README.md:
```diff
@@ -1,3 +1,135 @@
+    - [x] X축 (0-8), Y축(0-9)로 구현한다.
+- [x] 기물은 팀을 갖는다.
+    - [x] 팀은 한나라, 초나라로 나뉘어 진다.
+- [x] 기물은 움직인다.
+    - [x] 다른 기물과 상관없이 갈 수 있는 경로를 얻는다.
+- [x] 졸
+    - [x] 졸 기물은 좌, 우로 움직일 수 있고 한나라일 경우 하, 초나라일 경우 상으로 움직일 수 있다.
+- [x] 마
+    - [x] 마 기물은 직진 -> 직진 방향의 대각선 한 칸으로 움직인다. (ex 오른쪽 -> 오른쪽 위)
+- [x] 사
+    - [x] 사 기물은 상, 하, 좌, 우로 움직일 수 있다.
+- [x] 왕
+    - [x] 왕 기물은 상, 하, 좌, 우로 움직일 수 있다.
+- [x] 상
+    - [x] 상 기물은 직진 -> 직진 방향의 대각선 두 칸으로 움직인다. (ex 오른쪽 -> 오른쪽 위 -> 오른쪽 위)
+- [x] 차
+    - [x] 차 기물은 상, 하, 좌, 우 모두 원하는 만큼 움직일 수 있다.
+- [x] 포
+    - [x] 포 기물은 상, 하, 좌, 우 모두 원하는 만큼 움직일 수 있다.
+    - [x] 포 기물은 다른 포 기물을 제외한 다른 기물을 넘어 뛰어야한다.
```
기물별 자세한 이동 규칙을 정리해주셨네요 👍 👍 👍 
새로 합류한 사람이 장기에 대해 잘 몰라도 이해할 수 있겠어요 💯

Gomding commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,116 @@
+
+public class Cannon extends Piece {
+
+    private static final int REQUIRED_JUMP_PIECES = 1;
```
상수로 표현 👍 👍 👍

Gomding commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,116 @@
+        return isValidCannonRoute(route, otherPieces);
+    }
+
+    private boolean isValidCannonRoute(final Route route, final List<Piece> otherPieces) {
+        final List<Piece> piecesInRoute = otherPieces.stream()
+                .filter(piece -> route.hasPosition(piece) && !piece.isSamePosition(position))
+                .toList();
+
+        if (piecesInRoute.stream().anyMatch(Piece::isCannon)) {
+            return false;
+        }
+
+        final Position destination = route.getDestination();
+
+        final List<Piece> jumpPieces = new ArrayList<>();
+        Piece targetPiece = null;
+
+        for (final Piece piece : piecesInRoute) {
+            if (piece.isSamePosition(destination)) {
+                targetPiece = piece;
+                continue;
+            }
+            jumpPieces.add(piece);
+        }
+
+        if (jumpPieces.size() != REQUIRED_JUMP_PIECES) {
+            return false;
+        }
+
+        return targetPiece == null || !targetPiece.isSameTeam(team);
+    }
```
> 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.
>   - 함수(또는 메소드)가 한 가지 일만 하도록 최대한 작게 만들어라.

이전 미션의 규칙도 적용해보면 좋을 것 같아요!
적용하지 않는다면 이유를 말씀해주셔도 좋습니다 😃

CheChe903 replied:
[1ce6e20](https://github.com/woowacourse/java-janggi/pull/63/commits/1ce6e20b1b6e9edc54a7960ff6f1693b863b22a4)

최대한 수정해보았습니다.

Gomding commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,116 @@
+                break;
+            }
+        }
+        return positions;
```
> 이전 미션의 규칙도 적용해보면 좋을 것 같아요!
> 적용하지 않는다면 이유를 말씀해주셔도 좋습니다 😃

여기도 동일합니다! ㅎㅎ

Gomding commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,116 @@
+        return targetPiece == null || !targetPiece.isSameTeam(team);
+    }
+
+    @Override
+    public boolean isCannon() {
+        return true;
+    }
```
리팩토링 했으니 접근제어자를 변경해볼 수 있곘네요! 😃

CheChe903 replied:
[2870a8f](https://github.com/woowacourse/java-janggi/pull/63/commits/2870a8ff056a39e400917c726c5812de27877e16)

Gomding commented on src/main/java/janggi/domain/piece/direction/BoardSize.java:
```diff
@@ -0,0 +1,23 @@
+package janggi.domain.piece.direction;
+
+public enum BoardSize {
+
+    MIN_X(0),
+    MAX_X(8),
+    MIN_Y(0),
+    MAX_Y(9);
```
해당 값은 Position 내부에서 알고 있는것을 체체는 어떻게 생각하시나요?

Position 은 해당 값에 관계가 깊을 것 같아서요 :)
특히 해당 값과 관련해서 객체 생성 시 사용하고 있기도 하구요
BoardSize를 사용하는곳을 살펴보니 항상 Position 객체와 함께 사용하고 있군요 😄

CheChe903 replied:
저도 한 곳에서 관리하고 싶었지만,, 

`BoardSize`는 
```
    private int calculateMaxSteps(final Direction direction) {
        if (direction.dx() > 0) {
            return MAX_X.getSize() - position.x();
        } else if (direction.dx() < 0) {
            return position.x() - MIN_X.getSize();
        } else if (direction.dy() > 0) {
            return MAX_Y.getSize() - position.y();
        }
        return position.y() - MIN_Y.getSize();
    }
```
해당 `Cannon`, `Chariot`에 있는 메서드들에도 쓰여서 하나로 합치긴 어려울 것 같다고 판단해서 이렇게 구성했어요.

근데 리팩토링을 하다보니 없애는 방향으로도 가능하더라구요~! 그렇게 해보았어요.

[b78aac6](https://github.com/woowacourse/java-janggi/pull/63/commits/b78aac6db3284cc205c150b1d872aa4359976385)

Gomding commented on src/main/java/janggi/domain/piece/Elephant.java:
```diff
@@ -0,0 +1,67 @@
+
+public class Elephant extends Piece {
+
+    private static final List<List<Direction>> ELEPHANT_MOVES = List.of(
+            List.of(UP, RIGHT_UP, RIGHT_UP),
+            List.of(UP, LEFT_UP, LEFT_UP),
+
+            List.of(RIGHT, RIGHT_UP, RIGHT_UP),
+            List.of(RIGHT, RIGHT_DOWN, RIGHT_DOWN),
+
+            List.of(DOWN, LEFT_DOWN, LEFT_DOWN),
+            List.of(DOWN, RIGHT_DOWN, RIGHT_DOWN),
+
+            List.of(LEFT, LEFT_UP, LEFT_UP),
+            List.of(LEFT, LEFT_DOWN, LEFT_DOWN)
+    );
```
이해하기 좋게 구현해주셨네요 💯 💯 💯

CheChe903 replied:
감사합니다 ~!

Gomding reviewed (changes requested):
안녕하세요 체체!
리팩토링 깔끔하게 해주셨네요 👍 👍 
추가로 몇가지 코멘트 남겼으니 확인부탁드려요 :)

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요~

CheChe903 committed:
refactor(BoardSize): 보드 사이즈 삭제

- Position에 해당 상수들을 옮기고, Position에서 해결하도록 수정합니다.

CheChe903 committed:
fix: UP,DOWN 반대로 된 것 수정

CheChe903 committed:
test(Position): 포지션 단위 테스트 진행

CheChe903 committed:
test(Piece): 단위 테스트 수정

CheChe903 committed:
feat(Controller): 예외 처리 및 재입력 구현

CheChe903 committed:
refactor: 메서드 분리

CheChe903 committed:
refactor(Piece): isCannon 메서드 접근 제어자 변경

CheChe903 committed:
feat(Route): 가능한 경로만 입력받도록 설정

Gomding commented on src/test/java/janggi/domain/piece/CannonTest.java:
```diff
@@ -1,72 +1,125 @@
+                new Position(4, 0))));
+    }
+
     @DisplayName("포 기물이 올바른 루트를 계산하는지 확인한다.")
```
테스트를 세부사항으로 나눠서 테스트하라는것은 다음과 같은 사항들 이었어요
아래 내용을 읽어보면 요구사항 문서를 읽는 느낌이 들까요?

- 포는 상하좌우 이동이 가능하다
- 포 기물은 다른 포 기물을 제외한 다른 기물을 넘어 뛰어야한다.
- 포 기물은 다른 기물이 없으면 이동이 불가능하다
- 같은팀 기물은 공격이 불가능하다
- 포 기물은 포 기물에 대한 공격이 불가능하다
- 포 위로는 점프가 불가능하다

꼭 메서드와 1:1 로 테스트를 작성할 필요는 없어요. 
테스트로 메서드 내부의 작은 케이스 하나씩 테스트하는것이 체체가 말한 문제점을 찾는데 시간이 덜 걸리는 방향이라 생각해요

> 개발자가 스스로 실행(Application 실행)하며 테스트를 진행, 기능이 의도한 바대로 되지 않을 때 보통 테스트를 모두 돌려볼 것입니다. Board에 대한 테스트가 없어 테스트가 모두 성공한다면, 문제점을 찾는데 시간이 더 걸릴 것 같아요.

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -0,0 +1,75 @@
+    public void run() {
+        final Board board = generateBoard();
+        final List<Piece> pieces = board.getPieces();
+        while (true) {
```
장기에는 게임 종료 조건이 있을거에요!
중요한 규칙이라 생각해서 해당 부분도 꼭 구현해주세요 :)

Gomding commented on src/main/java/janggi/domain/piece/General.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public Set<Route> calculateIndependentRoutes() {
+        return GENERAL_MOVES.stream()
+                .map(this::calculateRoute)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toSet());
```
Optional 을 활용해주셨군요
아래 글의 3번 항목을 읽고 체체의 의견을 공유해주면 좋겠어요~

- [Java Optional 바르게 쓰기](https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)

Gomding commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,90 @@
+        return validRoutes;
+    }
+
+    private Set<Route> generateRoutesInDirection(final Direction direction) {
+        final Set<Route> directionalRoutes = new HashSet<>();
+
+        Position currentPosition = position;
+        final List<Position> positions = new ArrayList<>();
+        while (currentPosition.canMove(direction)) {
+            final Position nextPosition = currentPosition.move(direction);
+            positions.add(nextPosition);
+            directionalRoutes.add(new Route(new ArrayList<>(positions)));
+            currentPosition = nextPosition;
+        }
+        return directionalRoutes;
+    }
```
Chariot 과 공통된 부분으로 보이네요 :)
추상화 계층을 만들어서 중복된 부분을 추출해보는것은 어떻게 생각하시나요?

Gomding reviewed (approved):
안녕하세요 체체!
전체적으로 리팩토링 잘 해주셨네요 👍 👍 👍 
1단계는 이만 머지하겠습니다~
추가로 남긴 코멘트는 2단계 시작하면서 함께 고민해주세요!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요 😃
